tests/func/utils/test_strict_yaml.py

```diff
@@ -0,0 +1,291 @@
+"""
+
+DUPLICATE_KEYS_OUTPUT = """\
+'./dvc.yaml' is corrupted.
+
+While constructing a mapping, in line 3, column 5
+  3 │   cmd: python train.py
+
+Found duplicate key "cmd" with value "python train.py" (original value:\
+ "python \ntrain.py"), in line 4, column 5
+  4 │   cmd: python train.py"""
+
+
+MAPPING_VALUES_NOT_ALLOWED = """\
+stages:
+  stage1
+    cmd: python script.py
+"""
+
+MAPPING_VALUES_NOT_ALLOWED_OUTPUT = """\
+'./dvc.yaml' is corrupted.
+
+Mapping values are not allowed in this context, in line 3, column 8
+  3 │   cmd: python script.py"""
+
+
+NO_HYPEN_INDICATOR_IN_BLOCK = """\
+stages:
+  stage1:
+    cmd: python script.py
+    outs:
+      - logs:
+          cache: false
+      metrics:
+"""
+
+NO_HYPEN_INDICATOR_IN_BLOCK_OUTPUT = """\
+'./dvc.yaml' is corrupted.
+
+While parsing a block collection, in line 5, column 7
+  5 │     - logs:
+
+Did not find expected '-' indicator, in line 7, column 7
+  7 │     metrics:"""
+
+
+UNCLOSED_SCALAR = """\
+stages:
+  stage1:
+    cmd: python script.py
+    desc: "this is my stage one
+"""
+
+UNCLOSED_SCALAR_OUTPUT = """\
+'./dvc.yaml' is corrupted.
+
+While scanning a quoted scalar, in line 4, column 11
+  4 │   desc: "this is my stage one
+
+Found unexpected end of stream, in line 5, column 1
+  5"""
+
+
+NOT_A_DICT = "3"
+NOT_A_DICT_OUTPUT = "'./dvc.yaml' validation failed: expected a dictionary.\n"
+
+
+MISSING_CMD = """\
+stages:
+  stage1:
+"""
+
+MISSING_CMD_OUTPUT = """\
+'./dvc.yaml' validation failed.
+
+expected a dictionary, in stages -> stage1, line 2, column 3
+  1 stages:
+  2   stage1:
+  3"""
+
+
+DEPS_AS_DICT = """\
+stages:
+  stage1:
+    cmd: python script.py
+    deps:
+      - src:
+"""
+
+DEPS_AS_DICT_OUTPUT = """\
+'./dvc.yaml' validation failed.
+
+expected str, in stages -> stage1 -> deps -> 0, line 5, column 9
+  4 │   deps:
+  5 │     - src:
+"""
+
+NULL_VALUE_ON_OUTS = """\
+stages:
+  cmd: python script.py
+  outs:
+  - logs:
+      cache: false
+      persist: true
+      remote:
+"""
+
+NULL_VALUE_ON_OUTS_OUTPUT = """\
+'./dvc.yaml' validation failed: 2 errors.
+
+expected a dictionary, in stages -> cmd, line 2, column 3
+  1 stages:
+  2   cmd: python script.py
+  3   outs:
+
+expected a dictionary, in stages -> outs, line 4, column 3
+  3   outs:
+  4   - logs:"""
+
+ADDITIONAL_KEY_ON_OUTS = """\
+stages:
+  cmd: python script.py
+  outs:
+  - logs:
+      cache: false
+      not_existing_key: false
+"""
+
+ADDITIONAL_KEY_ON_OUTS_OUTPUT = """\
+'./dvc.yaml' validation failed: 2 errors.
+
+expected a dictionary, in stages -> cmd, line 2, column 3
+  1 stages:
+  2   cmd: python script.py
+  3   outs:
+
+expected a dictionary, in stages -> outs, line 4, column 3
+  3   outs:
+  4   - logs:"""
+
+
+FOREACH_SCALAR_VALUE = """\
+stages:
+  group:
+    foreach: 3
+    do:
+      cmd: python script${i}.py
+"""
+
+FOREACH_SCALAR_VALUE_OUTPUT = """\
+'./dvc.yaml' validation failed.
+
+expected dict, in stages -> group -> foreach, line 3, column 5
```

Comment:
Should have been something like following, but we have issues with validating "EitherOr" schema (at least it starts throwing too many errors in those cases).
```console
expected dict and list, in stages -> group -> foreach, line 3, column 5
```

Reply:
Not ideal, but at least it points to the right place!

Reply:
I have plans to work on that in the near future. :)